packages/workspace/src/browser/workspace-service.ts

```diff
@@ -88,78 +86,107 @@ export class WorkspaceService implements FrontendApplicationContribution {
-        this._workspaceFolder = await this.toValidRoot(rootUri);
-        const configUri = this.getWorkspaceConfigFileUri();
-        if (configUri) {
-            this.watcher.watchFileChanges(configUri);
+        if (mostRecentWorkspace && !mostRecentWorkspace.isDirectory) {
+            this._workspaceFile = mostRecentWorkspace;
+            this.workspaceFileWatcher = await this.watcher.watchFileChanges(new URI(this._workspaceFile.uri));
         }
     }
 
-    protected async updateRoots(): Promise<void> {
-        this._roots = await this.computeRoots();
+    protected async updateRoots(mostRecentWorkspace: FileStat | undefined): Promise<void> {
+        this._roots = await this.computeRoots(mostRecentWorkspace);
         this.deferredRoots.resolve(this._roots); // in order to resolve first
         this.deferredRoots = new Deferred<FileStat[]>();
         this.deferredRoots.resolve(this._roots);
         this.onWorkspaceChangeEmitter.fire(this._roots);
     }
 
-    protected async computeRoots(): Promise<FileStat[]> {
-        const config = await this.getWorkspaceConfig();
-        if (!config || !config.roots.length) {
-            return this._workspaceFolder ? [this._workspaceFolder] : [];
-        }
+    protected async computeRoots(workspaceFile: FileStat | undefined): Promise<FileStat[]> {
         const roots: FileStat[] = [];
-        for (const rootUri of config.roots) {
-            const root = await this.toValidRoot(rootUri);
-            if (root) {
-                roots.push(root);
+        if (workspaceFile) {
+            if (workspaceFile.isDirectory) {
+                return [workspaceFile];
             }
-        }
-        return roots;
-    }
 
-    protected async getWorkspaceConfig(): Promise<{ stat: FileStat, roots: string[] } | undefined> {
-        if (!this.preferences['workspace.supportMultiRootWorkspace']) {
-            return undefined;
-        }
-        const configUri = this.getWorkspaceConfigFileUri();
-        if (configUri) {
-            const uriStr = configUri.path.toString();
-            if (await this.fileSystem.exists(uriStr)) {
-                const { stat, content } = await this.fileSystem.resolveContent(uriStr);
-                if (content) {
-                    // FIXME use jsonc + ajx to parse & validate content
-                    const roots: string[] = JSON.parse(content).roots || [];
-                    return { stat, roots };
+            const workspaceData = await this.getWorkspaceDataFromFile(workspaceFile);
+            if (workspaceData) {
+                const invalidRoots: string[] = [];
+                for (const root of workspaceData.folders) {
+                    const valid = await this.toValidRoot(root.path);
+                    if (valid) {
+                        roots.push(valid);
+                    } else {
+                        invalidRoots.push(root.path);
+                    }
+                }
+                if (invalidRoots.length > 0) {
```

Comment:
why do we need to throw notification to a user about such issues? the root gets deleted, it is not part of the workspace. Is it a big deal to show notifications?

Reply:
this line of code would be hit if a folder is moved, deleted, or renamed. i am notifying the users because:

- unlike internal errors such as "undefined is not an object" in the log, this is not a problem with the product itself. Instead, it is most likely caused by users, or other software applications running in the same environment. in my opinion, whoever caused the error should see message, and perform the fix. 
- users would be confused if s/he has less roots opened than s/he has in the workspace file. e..g., how come theia only opens 4 roots while I have 6 in the workspace, what's happening?
- it is true that users could figure out what happened from the logs. however they don't always have access to logs. e.g., i don't know where to find gitpod logs, and AFAIK, Gaia users in Ericsson couldn't see the logs either. 

If you really dislike the notification, maybe we could think about doing what @marcdumais-work proposed:

>If the workspace file is not `WorkspaceData` or paths are invalid, we open the workspace file in an editor, and add error markers.

Please let me know what you think. Thank you !

Reply:
I think it is a corner case. Using multi-root workspaces with VS Code for one year, it never happened to me. We could wait till it turns in the real issue and then come up with a solution.

> If the workspace file is not WorkspaceData or paths are invalid, we open the workspace file in an editor, and add error markers.

I don't think a user supposed to see this file, it is an implementation detail. Also, it is subject to change: as you remember discussing with @svenefftinge it should be configurable in the future what is a source of workspace, not necessary has to be a file.

As a solution, I would imagine to always show workspace roots in the navigator to let a user deleting them via UI.

Reply:
@akosyakov The latest code in this PR follows your suggestion:

All workspace roots, including the invalid, are displayed in the navigator. Users would be able to tell the which ones are invalid from the warning overlay rendered on the bottom-right corner of the folder icon.